python/formatting.rst

```diff
@@ -0,0 +1,107 @@
+.. _formatting-python-code:
+
+####################################
+Formatting Python Code Automatically
+####################################
+
+The Python ecosystem provides two packages that can simplify automated code formatting and are used by some software products.
+These tools are compliant with the :ref:`Rubin style guide <style-guide-py-pep8-baseline>` and can be adopted for any repository.
+The main caveat is that once this tooling is adopted for a specific repository, use of the tooling must be required for all subsequent pull requests and checked using GitHub Actions.
+This ensures that the code only goes through a style change for the initial adoption of the tooling.
+
+Ordering imports
+================
+
+The `isort <https://pycqa.github.io/isort/>`_ package can be used to reorder imports in your Python code such that they are in alphabetical order and distinguish core python packages from other packages.
+
+.. code-block:: bash
+
+    $ isort python
+
+Formatting Python Code
+======================
+
+We recommend the use of the `black <https://black.readthedocs.io/en/stable/index.html>`_ command for reformatting Rubin Python code once it has been decided that a repository should be converted.
+Black calls itself the "uncompromising code formatter" and its entire purpose is to remove discussions of style from the debate.
+You can write the code however you want in your editor but when ``black`` runs on it there is no room for debate.
+For this reason configuration options are limited.
+The only configuration option that is recommended is to control the line length to meet the Rubin coding style.
+
+``black`` is not compatible with the E203 error from Flake8.
+Consequently any project using ``black`` should disable E203.
+
+Both ``black`` and ``isort`` can be configured by using a file ``pyproject.toml`` in the repository root:
+
+.. literalinclude:: examples/pyproject.toml
+  :language: toml
+
+The only configuration sets the line length and Python version, with ``isort`` configured to be compatible with ``black``.
+
+Installation
+============
+
+``isort`` and ``black`` are both available from conda-forge and can be installed directly into a conda environment:
+
+.. code-block:: bash
+
+    $ mamba-install black isort
+
+Of course they can also be installed using ``pip``.
+They are not currently part of the default ``rubin-env`` conda environment.
+
+Once installed it is possible to configure editors to automatically apply these tools on save.
```

Comment:
I didn't add examples because I don't know how to do it. I use `pre-commit` mostly. I know @TallJimbo does it so maybe he can add something for vscode when he is done. I don't know if the editor can do it on a per-repository basis since you don't want it to run black on code that is not yet using it.